### ISSUE TITLE
Type inference can't incorrectly inferes possibly nullable type argument from non-nullable arguments

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#49295]
**Version Used**: 
```
Branch master (10 Nov 2020)
Latest commit ec71e62 by CyrusNajmabadi:
Merge pull request #49242 from CyrusNajmabadi/unconstrainedImplInterface

Do not convert an unconstrained generic to a nullable struct.
```

**Steps to Reproduce**:

Compile the following code
```cs
#nullable enable
class C
{
  void M1<TEnumerator>(TEnumerator e) where TEnumerator: new()
  {
    if (e == null) return;

    e = new();
    e.ToString();
    var enumerable = Id(e);
    enumerable.ToString();
  }

  void M2<TEnumerator>(TEnumerator e) where TEnumerator: new()
  {
    e = new();
    e.ToString();
    var enumerable = Id(e);
    enumerable.ToString();
  }

  static T Id<T>(T t) => t;
}
```

**Expected Behavior**:
No warnings. The variables `e` and `enumerable` are definitely not nullable in both mehod

**Actual Behavior**:
`CS8602: Dereference of a possibly null reference.` is reported for `enumerable.ToString()` in both methods

**Comment by jaredpar** [comment]
This behavior is "By Design". 

In this case `e` is indeed *nonnullable* because we can trace the use of `new`. That does not necessarily mean though that `TEnumerator` itself is a *nonnullable* type. This could be legally instantiated with say `string?`. Given that the nullable state of of `Id<TEnumerator>` is *maybe default* and we properly report a warning when `ToString` is invoked here.

**Comment by TessenR** [comment]
@jaredpar 
Could you please share the goals / motivation for this design? I couldn't come up with an example where this behavior is required / beneficial
I see 2 problems that led me to conclusion that this might be a bug

First: there is no way for a method with signature like `T Id<T>(T t)` to return a _maybe default_ value unless its argument is _maybe default_. The reason for this is that the compiler will warn on `return default` or any other source of _maybe default_ value since `T` might be not nullable. It will only allow returns of non-nullable values e.g. `new T()` or the input parameter since it already satisfies the required nullability. Hence if the argument has _nonnullable_ state the method cannot return _maybedefault_ and the warning is a false positive.

Second: all other types except for unconstrained generics and `Nullable<T>` contribute their _state_ instead of their _annotation_ for type inference.
e.g.
```
string? nullableStr = "";
var notNullableStr = Id(nullableStr);
notNullableStr.ToString(); // no problems here

T Id<T>(T t) => t;
```
I believe this example can as well be used in your argument like this: In this case nullableStr is indeed nonnullable because the compiler can trace its initializer. However this does not mean that string? is a nonnullable type.
Why these two cases are handled differently then?

These two points make this case look like a compiler bug from the user's perspective. 
I might be speculating there as I didn't check the compiler's sources very thoroughly but it looks like the compiler needs `TypeWithAnnotation` to perform type inference so it tries to map the argument's `TypeWithState` to `TypeWithAnnotation` and while NotAnnotated annotation correctly expresses non-nullable string type, there's no `NullableAnnotation` enum member to express non-nullable unconstrained generic type so the result state is accidentally coalesced to _maybedefault_ due to this unrepresentable annotation. This made me wonder whether it's indeed a design choice based on some example where the warning is correct and useful or is it just a side-effect of such mapping that results in a false-positive warning that could be treated as a bug
